[ADD] hr_personal_equipment: added module to manage employee personal equipment

albariera opened the pull request:
This module allows managing employee personal equipment by creating personal equipment requests and controlling material allocations. 
We are currently working on the next modules that could be integrated with this one: 
- hr_personal_equipment_stock: This module will integrate the personal equipment request with stock in order to have better equipment control if necessary. When the request is accepted, a stock request will be generated.
- hr_personal_equipment_variant_configurator: This module will integrate the module "product_variant_configurator" for being able to choose between different product variants when selecting personal equipment. 
- hr_employee_ppe: This module will contain additional functionalities for managing personal protection equipment, as adding equipment duration and receipt report. This module was already in OCA, but several changes have been done in order to integrate it with hr_personal_equipment. A migration script will also be added. 

Any idea of improvement will be welcome.

albariera committed:
[ADD] hr_personal_equipment: added module to manage employee personal equipment

etobella commented:
@marcelsavegnago That is what we talked. This module can be used as base for all the equipments for employees, including ppe

etobella commented on hr_personal_equipment/views/hr_personal_equipment_request.xml:
```diff
@@ -0,0 +1,126 @@
+    </record>
+
+    <record model="ir.ui.menu" id="hr_personal_equipment_request_menu">
+        <field name="name">Employee Personal Equipment Request</field>
```
Can you make this shorter?

albariera replied:
Done! Thank you for the review!

yann-papouin commented on hr_personal_equipment/views/hr_personal_equipment.xml:
```diff
@@ -0,0 +1,116 @@
+                <field name="employee_id" groups="hr.group_hr_user"/>
+                <field name="product_id"/>
+                <group expand="0" name="group_by" string="Group By">
+                    <filter name="state" string="State" context="{'group_by' : 'state'}" />
+                    <filter name="product_id" string="Product" context="{'group_by' : 'product_id'}" />
+                    <filter name="employee_id" string="Employee" context="{'group_by' : 'employee_id'}" />
+                </group>
+                <filter string="In Draft" name="draft_state" domain="[('state', '=', 'draft')]"/>
+                <filter string="Accepted" name="accepted_state" domain="[('state', '=', 'accepted')]"/>
+                 <filter string="Valid" name="valid_state" domain="[('state', '=', 'valid')]"/>
+                <filter string="Expired" name="expired_state" domain="[('state', '=', 'expired')]"/>
```
Please never use a name for a filter that already exists as a field or it will collide when set from a context using `default_`.

My advice, always add `filter_` as a prefix or `groupby_`:
```
<group expand="0" name="group_by" string="Group By">
  <filter name="groupby_product_id" string="Product" context="{'group_by' : 'product_id'}" />
</group>
<filter string="In Draft" name="filter_draft_state" domain="[('state', '=', 'draft')]"/>
```

albariera replied:
Done! Thank you for your comment!

albariera commented:
@OCA/human-resources-maintainers Could you give us some help, please?
The idea would be to do a refactor of the current OCA module hr_employee_ppe in order to have a more general module to manage employee material. So, with this refactor, the hr_employee_ppe would depend on hr_employee_material. 
@marcelsavegnago is the author of the original hr_employee_ppe and after talking privately with him, he has agreed to do this refactor. 
It would be great if you could provide some feedback. Thank you! :smile:

Saran440 commented on hr_personal_equipment/__manifest__.py:
```diff
@@ -0,0 +1,21 @@
+    'version': '12.0.1.0.0',
+    'license': 'AGPL-3',
+    'author': 'Creu Blanca,Odoo Community Association (OCA)',
+    'website': 'www.creublanca.es',
```
```suggestion
    'website': 'https://github.com/OCA/hr',
```

Saran440 commented on hr_personal_equipment/models/hr_personal_equipment.py:
```diff
@@ -0,0 +1,70 @@
+    def _compute_name(self):
+        for rec in self:
+            if rec.product_id.name and rec.employee_id.name:
+                rec.name = rec.product_id.name + _(" to ") + rec.employee_id.name
```
```suggestion
                rec.name = "{} to {}".format(rec.product_id.name, rec.employee_id.name)
```

I suggest use {} instead of + for python3

reference : https://realpython.com/python-string-formatting/

Saran440 commented on hr_personal_equipment/models/hr_personal_equipment.py:
```diff
@@ -0,0 +1,70 @@
+    def _validate_allocation_vals(self):
+        return {
+            'state': 'valid',
+            'start_date': date.today() if not self.start_date else self.start_date,
```
```suggestion
            'start_date': fields.Date.context_today(self) if not self.start_date else self.start_date,
```

`fields.Date.context_today(self)` support datetime with timezone.
I'm not sure that `date.today()` support with timezone.
If I live in Thailand (GMT +7), Is it correct?

Saran440 commented on hr_personal_equipment/readme/CREDITS.rst:
```diff
@@ -0,0 +1 @@
+
```
if you not used, please delete.

Saran440 commented on hr_personal_equipment/readme/INSTALL.rst:
```diff
@@ -0,0 +1 @@
+To install this module, you need to have HR module installed or it will be requested during installation.
```
it's not necessary to explain. please delete this file.

Saran440 reviewed (commented):
Code Review.

albariera commented:
@Saran440 Comments attended! Thank you for the review :smiley:

albariera commented:
@Saran440 Could you re-review, please? :blush:

Saran440 reviewed (commented):
@albariera Functional Test no bug.

1. May be should this module name `hr_employee_equipment_request`?
2. After employee request equipment, Can you add smart button on each employee  link to equipment? (Not sure that makes sense)

albariera commented:
Thank you for your comments @Saran440. I agree with you. 

I have changed the module name and added two smart buttons on the employee form: one for personal equipment requests and the other for personal equipment allocations. I have named the former "PE Request" and the latter "PE", because the other names were very long and ended with "...". If you come up with a better name please let me know.

I have also put as an icon a briefcase, I would have liked a t-shirt, but they are not available in font awesome v.4. So if it also occurs to you a better icon let me know too.

I have also added tests for these buttons.

Thank you again!

AaronHForgeFlow reviewed (approved):
Code review LGTM!

dreispt commented:
/ocabot merge nobump
